scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -160,41 +189,113 @@
-                                 [:missing :pod/failed] (remove-finalization-if-set-and-delete api-client expected-state-dict pod)
+                                 ; There was a race and it completed normally before being it was killed.
+                                 [:expected/killed :pod/succeeded] (pod-has-just-completed kcc existing-state-dict)
+                                 [:expected/killed :pod/failed] (pod-has-just-completed kcc existing-state-dict)
+                                 [:expected/killed :missing] (do ; Weird. We always update datomic first. Could happen if someone manually removed stuff from kubernetes.
+                                                               (log-weird-state expected-state-dict existing-state-dict)
+                                                               (pod-was-killed kcc pod-name))
+
+                                 ; These cases are weird.
+                                 [:expected/completed :pod/waiting] (kill-task-weird api-client expected-state-dict existing-state-dict)
+                                 [:expected/completed :pod/running] (kill-task-weird api-client expected-state-dict existing-state-dict)
+
+                                 ; This is interesting. This indicates that something deleted it behind our back!
+                                 ; Other 3 cases of [{completed,starting,killed}, :missing] already handled.
+                                 [:expected/running :missing] (do
+                                                                (log/error "Something deleted" pod-name "behind our back")
+                                                                (pod-has-just-completed kcc existing-state-dict)) ; TODO: Should mark mea culpa retry
+
+                                 ; This block of :missing,* cases can only occur in testing when you're e.g., blowing away the database. 
+                                 ; The next two will go through :missing,running or :missing,waiting to :missing,failed and then be deleted.
+                                 ; TODO: These may be evidence of a bug where we process pod changes when we're starting up.
+                                 [:missing :pod/running] (kill-task-weird api-client nil pod)
+                                 [:missing :pod/waiting] (kill-task-weird api-client nil pod)
+
+                                 ; We shouldn't hit these unless we get a database rollback.
+                                 [:missing :pod/succeeded] (kill-task-weird api-client nil pod)
+                                 [:missing :pod/failed] (kill-task-weird api-client nil pod)
+
+                                 [:missing :pod/unknown] (kill-task-weird api-client expected-state-dict existing-state-dict)
+                                 [:expected/starting :pod/unknown] ; TODO: Should mark mea culpa retry
```

Comment:
Nitpick: Let's stick to the state pair on the left and action on the right on the same line scheme.

Reply:
Ok.